src/editor/input_center.cpp

```diff
@@ -0,0 +1,738 @@
+#include "supertux/game_object_ptr.hpp"
+#include "supertux/level.hpp"
+#include "supertux/moving_object.hpp"
+#include "supertux/object_factory.hpp"
+#include "supertux/sector.hpp"
+#include "supertux/tile.hpp"
+#include "supertux/tile_manager.hpp"
+#include "supertux/tile_set.hpp"
+#include "trigger/climbable.hpp"
+#include "trigger/scripttrigger.hpp"
+#include "trigger/secretarea_trigger.hpp"
+#include "trigger/sequence_trigger.hpp"
+#include "video/font.hpp"
+#include "video/renderer.hpp"
+#include "video/video_system.hpp"
+#include "math/vector.hpp"
+
+EditorInputCenter::EditorInputCenter() :
+  hovered_tile(0, 0),
+  sector_pos(0, 0),
+  mouse_pos(0, 0),
+  dragging(false),
+  drag_start(0, 0),
+  dragged_object(NULL),
+  marked_object(NULL),
+  edited_path(NULL),
+  object_tip(),
+  obj_mouse_desync(0, 0),
+  render_grid(true),
+  scrolling(Vector())
+{
+}
+
+EditorInputCenter::~EditorInputCenter()
+{
+}
+
+void
+EditorInputCenter::update(float elapsed_time) {
+  update_scroll();
+}
+
+void
+EditorInputCenter::delete_markers() {
+  for (auto i = Editor::current()->currentsector->moving_objects.begin();
+       i != Editor::current()->currentsector->moving_objects.end(); ++i) {
+    MovingObject* moving_object = *i;
+    PointMarker* marker = dynamic_cast<PointMarker*>(moving_object);
+    if (marker) {
+      marker->remove_me();
+    }
+  }
+  marked_object = NULL;
+  edited_path = NULL;
+  Editor::current()->currentsector->update(0);
+}
+
+Rectf
+EditorInputCenter::drag_rect() {
+  int start_x, start_y, end_x, end_y;
+
+  if (drag_start.x < sector_pos.x) {
+    start_x = drag_start.x;
+    end_x = sector_pos.x;
+  } else {
+    start_x = sector_pos.x;
+    end_x = drag_start.x;
+  }
+
+  if (drag_start.y < sector_pos.y) {
+    start_y = drag_start.y;
+    end_y = sector_pos.y;
+  } else {
+    start_y = sector_pos.y;
+    end_y = drag_start.y;
+  }
+
+  return Rectf( Vector(start_x, start_y), Vector(end_x, end_y) );
+}
+
+void
+EditorInputCenter::input_tile(Vector pos) {
+  if ( !Editor::current()->layerselect.selected_tilemap ) {
+    return;
+  }
+
+  if ( pos.x < 0 || pos.y < 0 ||
+       pos.x >= ((TileMap *)Editor::current()->layerselect.selected_tilemap)->get_width() ||
+       pos.y >= ((TileMap *)Editor::current()->layerselect.selected_tilemap)->get_height()) {
+    return;
+  }
+
+  ((TileMap *)Editor::current()->layerselect.selected_tilemap)->change(pos.x, pos.y,
+                                                                       Editor::current()->tileselect.tile);
+}
+
+void
+EditorInputCenter::put_tile() {
+  input_tile(hovered_tile);
+}
+
+void
+EditorInputCenter::draw_rectangle() {
+
+  Rectf dr = drag_rect();
+  dr.p1 = sp_to_tp(dr.p1);
+  dr.p2 = sp_to_tp(dr.p2);
+
+  for (int x = dr.p1.x; x <= dr.p2.x; x++) {
+    for (int y = dr.p1.y; y <= dr.p2.y; y++) {
+      input_tile( Vector(x,y) );
+    }
+  }
+}
+
+void
+EditorInputCenter::fill() {
+
+  if (! Editor::current()->layerselect.selected_tilemap) {
```

Comment:
Please introduce a variable for the selected tilemap:
auto selected_tilemap = Editor::current()->layerselect.selected_tilemap

and replace all the following occurrences of the lengthy code with that.